README_ENGINE.md

```diff
@@ -37,55 +45,204 @@ The engine source code can be found in the `/engine` folder. The source code is
 
-## No C++ Exceptions
 We don’t make use of any exceptions in the engine. It isn’t generally used in game engines, since the data is (mostly) known beforehand, during development. Removing the support for C++ exceptions decreases executable size and improves the runtime performance.
 
+#### No STL
+
+While it might be tempting to use `std` functions/types/classes, we don't do it for several reasons.
+Performance is one (code size, compilation time, debugging time), and ABI issues is another important issue.
+
+You can read abit more about the advice we give to extension developers here:
+https://defold.com/manuals/extensions-best-practices/#standard-template-libraries---stl
+
+We do make a (very) few exceptions, most notable for `std::sort()` and `std::upper_bound()`.
+Although eventually, these may be replaced too.
+
+#### Memory allocations
+
+Most often we try to allocate memory upfront, when loading resources (e.g. loading a `.collection` file).
+We try to avoid runtime allocations as much as possible.
+
+We often use `free`, `malloc` and `realloc` to emphasize that the structs doesn't need a constructor.
+
+#### Avoid defensive programming
+
+Avoid adding defensive code such as `if (ptr != 0) *ptr = ...` as much as possible.
+Checks for validity should be done as early as possible. Best option would be in the editor, then the build pipeline, the resource loading in the engine, then the Lua interface. After that, the resource should either be ok, or the owner should handle it gracefully.
+
+Another example is setting memory to zero, just before explicitly setting the members one by one.
+In such a case, the memset isn't needed.
+
+Avoiding patterns like this keeps code size to a minimum.
+
+#### Optimizations
+
+When people mention optimizations, they can mean different things, and often they refer to a specific stage at the end of a project.
+We want performance to be there from day one, making sure each project can be developed as quickly as possible.
+
+We always try to think of the use cases and the expected performance in the design phase of a feature.
+We are not shy to implement features ourselves, in order to reach those goals.
+
+Also, when doing optimizations do measure them, and attach the info to the pull request, so the reviewers can make a proper assesment of the fix.
+E.g. compile time, runtime, memory etc.
+
+
+#### Third party libraries
+
+Sometimes we need to use a third party library.
+It is fine, if the cost of using it (code size, complexity, performance) is properly assessed.
+Also make sure to list alternatives (such as implementing the feature yourself) in the design phase.
+
+#### Debug vs Release
+
+We don't really distinguish between the traditional Debug or Release concept.
+We compile (and ship) all our libraries with the `-O2` flag.
+
+Instead, we compartmentalize functionality into separate libraries, and during link time, either add `featureX_impl.a` or a `featureX_null.a` to the command line.
```

Comment:
Should we say this is how we handle platform differences as well? feature_win32 feature_darwin? and #ifdefs if it's a small difference?